org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenImpl.java

```diff
@@ -714,56 +713,104 @@ private Properties loadLastUpdated(ArtifactRepository localRepository, Artifact
-        + artifact.getBaseVersion() + PATH_SEPARATOR;
+      ConfigurationConverter typeConverter = lookup.lookupConverterForType(clazz);
+      Object value = typeConverter.fromConfiguration(lookup, configuration, clazz, mojo.getImplementationClass(),
+          pluginRealm, evaluator, null);
+      return clazz.cast(value);
+    }
   }
 
-  private String formatAsDirectory(String directory) {
-    return directory.replace(GROUP_SEPARATOR, PATH_SEPARATOR);
+  private class ConfigurationElementImpl implements ConfigurationElement {
+    final PlexusConfiguration configuration;
+
+    final String name;
+
+    final ValueComputer valueComputer;
+
+    private ConfigurationElementImpl(PlexusConfiguration configuration, String name, ValueComputer valueComputer) {
+      this.name = name;
+      this.configuration = configuration;
+      this.valueComputer = valueComputer;
+    }
+
+    @Override
+    public boolean exists() {
+      return configuration != null;
+    }
+
+    @Override
+    public ConfigurationParameter get(String name) {
+      PlexusConfiguration child = configuration.getChild(name);
+      return new ConfigurationParameterImpl(child, this.name + "/" + name, valueComputer);
+    }
+
+    @Override
+    public Stream<ConfigurationParameter> children(String name) {
+      return Arrays.stream(configuration.getChildren(name))
+          .map(c -> new ConfigurationParameterImpl(c, this.name + "/" + name, valueComputer));
+    }
+
+    @Override
+    public Stream<ConfigurationParameter> children() throws NoSuchElementException {
+      return Arrays.stream(configuration.getChildren())
+          .map(c -> new ConfigurationParameterImpl(c, this.name + "/" + name, valueComputer));
+    }
+
   }
 
-  private <T> T getMojoParameterValue(MavenSession session, MojoExecution mojoExecution, String parameter,
-      Class<T> asType) throws CoreException {
-    try {
-      MojoDescriptor mojoDescriptor = mojoExecution.getMojoDescriptor();
+  private class ConfigurationParameterImpl extends ConfigurationElementImpl implements ConfigurationParameter {
 
-      ClassRealm pluginRealm = lookup(BuildPluginManager.class).getPluginRealm(session,
-          mojoDescriptor.getPluginDescriptor());
+    private ConfigurationParameterImpl(PlexusConfiguration configuration, String name, ValueComputer valueComputer) {
+      super(configuration, name, valueComputer);
+    }
 
-      ExpressionEvaluator expressionEvaluator = new PluginParameterExpressionEvaluator(session, mojoExecution);
-      ConfigurationConverter typeConverter = converterLookup.lookupConverterForType(asType);
-      Xpp3Dom dom = mojoExecution.getConfiguration();
-      if(dom == null) {
-        return null;
+    @Override
+    public <T> T as(Class<T> clazz) throws NoSuchElementException {
+      if(!exists()) {
+        throw new NoSuchElementException(
+            "Plugin execution " + valueComputer.mojo.getId() + "does not have a configuration parameter " + name);
```

Comment:
this should emit the full path to the parameter

Reply:
The name/path is already the full path. When a new `ConfigurationParameterImpl` is created the parents path concatenated with the elements name is passed as path.